### DISCUSSION TITLE
How do I use an older Keycloak client with Quarkus 2.13?

### DISCUSSION THREAD

**Discussion by vincejv** [quarkusio/quarkus#28649] [Q&A]
Hello all,

Would like to ask how to override the Keycloak version used by the Quarkus 2.13 BOM to Keycloak 18? My quarkus service could not connect to our Keycloak servers which still use v18. I have lost the error message in my logs but it said something about client secret is invalid when using Keycloak v19 client, but when downgrading to v18, there were no issues.

This PR https://github.com/quarkusio/quarkus/pull/27387 bumped the Keycloak version and it broke compatibility. I have tried overriding the property `<keycloak.version>18.0.1</keycloak.version>` in POM but it doesn't work, it keeps using the same dependency `org.keycloak:keycloak-core:19.0.2` and I even tried overriding the dependency `quarkus-keycloak-authorization` like so

```xml
<dependency>
  <groupId>io.quarkus</groupId>
  <artifactId>quarkus-keycloak-authorization</artifactId>
  <version>2.12.3.Final</version>  <!-- last quarkus version that used keycloak 18 -->
</dependency>
```

but it's still the same thing? I'm using a managed Keycloak server - Cloud IAM (Free tier) so any server upgrades wouldn't be easy.

[a comment by an automated account is not shown]

**Comment by vincejv** [comment]
Just ran a quick test, and it throws this error, so I presume Keycloak v19 client is not backwards compatible with v18 server
```java
2022-10-18 06:50:14,785 ERROR [io.qua.run.Application] (Quarkus Main Thread) Failed to start application (with profile prod): 
org.keycloak.authorization.client.util.HttpResponseException: Unexpected response from server: 
401 / Unauthorized / Response from server: {"error":"unauthorized_client","error_description":"Invalid client secret"}
	at org.keycloak.authorization.client.util.HttpMethod.execute(HttpMethod.java:95)
	at org.keycloak.authorization.client.util.HttpMethodResponse$2.execute(HttpMethodResponse.java:50)
```

**Comment by sberyozkin** [comment]
Hi @vincejv I've opened https://github.com/keycloak/keycloak/issues/14962, please watch it

**Reply by vincejv** [reply]
Hi @sberyozkin, I deleted my answer as it was wrong/did not do the intended thing. The code below overrode the other plugins/dependencies and etc not related to keycloak or authorizations including `cdi` and `arc` and still effectively using Quarkus `2.12.3.Final` instead of `2.13.2.Final`

```xml
<dependencyManagement>
    <dependencies>

      <!-- retain compatibility with keycloak 18 -->
      <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-keycloak-authorization</artifactId>
        <version>2.12.3.Final</version>  <!-- last quarkus version that used keycloak 18 -->
        <type>pom</type>
        <scope>import</scope>
      </dependency>

      <dependency>
        <groupId>${quarkus.platform.group-id}</groupId>
        <artifactId>${quarkus.platform.artifact-id}</artifactId>
        <version>2.13.2.Final</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>

  </dependencies>
</dependencyManagement>
```

Is there any other way to override the Keycloak Authz Client used besides downgrading the entire Quarkus BOM?

**Reply by sberyozkin** [reply]
May be only override `keycloak.version` to 18 ?

**Reply by vincejv** [reply]
> May be only override `keycloak.version` to 18 ?

you mean something like this?
```xml
<properties>
  <keycloak.version>18.0.2</keycloak.version>
</properties>
```
tried that, still used the KC Client v19
